packages/decoders/decoders.0.1.0/opam

```diff
@@ -0,0 +1,178 @@
+
+```ocaml
+utop # D.decode_value;;
+- : 'a decoder -> value -> ('a, error) result = <fun>
+```
+
+For convenience we also have `decode_string`, which takes a `string` and calls
+`Yojson`'s parser under the hood.
+
+```ocaml
+utop # D.decode_string D.int "1";;
+- : (int, error) result = Ok 1
+```
+
+What about a `list` of `int`s? Here's where the "combinator" part comes in.
+
+```ocaml
+utop # D.decode_string D.(list int) "[1,2,3]";;
+- : (int list, error) result = Ok [1; 2; 3]
+```
+
+Success!
+
+Ok, so what if we get some unexpected JSON?
+
+```ocaml
+utop # #install_printer D.pp_error;;
+utop # D.decode_string D.(list int) "[1,2,true]";;
+- : (int list, error) result =
+Error while decoding a list: element 2: Expected an int, but got true
+```
+
+## Generic decoders
+
+
+Suppose our program deals with users and roles. We want to decode our JSON input
+into these types.
+
+```ocaml
+type role = Admin | User
+
+type user =
+  { name : string
+  ; roles : role list
+  }
+```
+
+Let's define our decoders. We'll write a module functor so we can re-use the
+same decoders across different JSON libraries, with YAML input, or with
+Bucklescript.
+
+```ocaml
+module My_decoders(D : Decoders.Decode.S) = struct
+  open D
+  
+  let role : role decoder =
+    string >>= function
+    | "ADMIN" -> succeed Admin
+    | "USER" -> succeed User
+    | _ -> fail "Expected a role"
+    
+  let user : user decoder =
+    field "name" string >>= fun name ->
+    field "roles" (list role) >>= fun roles ->
+    succeed { name; roles }
+end
+
+module My_yojson_decoders = My_decoders(Decoders_yojson.Basic.Decode)
+```
+
+Great! Let's try them out.
+
+```ocaml
+utop # open My_yojson_decoders;;
+utop # D.decode_string role {| "USER" |};;
+- : (role, error) result = Ok User
+
+utop # D.decode_string D.(field "users" (list user))
+         {| {"users": [{"name": "Alice", "roles": ["ADMIN", "USER"]},
+                       {"name": "Bob", "roles": ["USER"]}]} 
+          |};;
+- : (user list, error) result =
+Ok [{name = "Alice"; roles = [Admin; User]}; {name = "Bob"; roles = [User]}]
+```
+
+Let's introduce an error in the JSON:
+
+```ocaml
+utop # D.decode_string D.(field "users" (list user))
+         {| {"users": [{"name": "Alice", "roles": ["ADMIN", "USER"]},
+                       {"name": "Bob", "roles": ["SUPER_USER"]}]}
+          |};;
+- : (user list, error) result =
+Error
+ in field "users":
+   while decoding a list:
+     element 1:
+       in field "roles":
+         while decoding a list:
+           element 0: Expected a role, but got "SUPER_USER"
+```
+
+We get a nice pointer that we forgot to handle the `SUPER_USER` role.
+
+## Release
+
+After updating CHANGES.md:
+
+```
+npm version <newversion>
+dune-release tag
+dune-release bistro -p decoders,decoders-ezjsonm,decoders-yojson
+```"""
+maintainer: "Matt Bray <matt@aestheticintegration.com>"
+authors: "Matt Bray <matt@aestheticintegration.com>"
+license: "LICENSE.md"
```

Comment:
mmh this is not what this field is about. It should state (roughly if you need to) what the name of the license si